crates/matrix-crypto-ffi/src/lib.rs

```diff
@@ -23,18 +23,308 @@ pub use device::Device;
+}
+
+/// A pickled version of an `InboundGroupSession`.
+///
+/// Holds all the information that needs to be stored in a database to restore
+/// an InboundGroupSession.
+#[derive(Debug, Deserialize, Serialize)]
+pub struct PickledInboundGroupSession {
+    /// The pickle string holding the InboundGroupSession.
+    pub pickle: String,
+    /// The public curve25519 key of the account that sent us the session
+    pub sender_key: String,
+    /// The public ed25519 key of the account that sent us the session.
+    pub signing_key: HashMap<String, String>,
+    /// The id of the room that the session is used in.
+    pub room_id: String,
+    /// The list of claimed ed25519 that forwarded us this key. Will be empty if
+    /// we directly received this session.
+    pub forwarding_chains: Vec<String>,
+    /// Flag remembering if the session was directly sent to us by the sender
+    /// or if it was imported.
+    pub imported: bool,
+    /// Flag remembering if the session has been backed up.
+    pub backed_up: bool,
+}
+
+/// Error type for the migration process.
+#[derive(thiserror::Error, Debug)]
+pub enum MigrationError {
+    /// Generic catch all error variant.
+    #[error("error migrating database: {message}")]
+    Generic {
+        /// The error message
+        message: String,
+    },
+}
+
+impl From<anyhow::Error> for MigrationError {
+    fn from(e: anyhow::Error) -> MigrationError {
+        MigrationError::Generic { message: e.to_string() }
+    }
+}
+
+/// Migrate a libolm based setup to a vodozemac based setup stored in a Sled
+/// store.
+///
+/// # Arguments
+///
+/// * `data` - The data that should be migrated over to the Sled store.
+///
+/// * `path` - The path where the Sled store should be created.
+///
+/// * `passphrase` - The passphrase that should be used to encrypt the data at
+/// rest in the Sled store. **Warning**, if no passphrase is given, the store
+/// and all its data will remain unencyrpted.
+///
+/// * `progress_listener` - A callback that can be used to introspect the
+/// progress of the migration.
+pub fn migrate(
+    mut data: MigrationData,
+    path: &str,
+    passphrase: Option<String>,
+    progress_listener: Box<dyn ProgressListener>,
+) -> anyhow::Result<()> {
+    use anyhow::Result;
+    use matrix_sdk_crypto::{
+        olm::PrivateCrossSigningIdentity,
+        store::{Changes as RustChanges, CryptoStore, RecoveryKey},
+    };
+    use matrix_sdk_sled::CryptoStore as SledStore;
+    use tokio::runtime::Runtime;
+    use vodozemac::{
+        megolm::InboundGroupSession,
+        olm::{Account, Session},
+        Curve25519PublicKey,
+    };
+    use zeroize::Zeroize;
+
+    // The total steps here include all the sessions/inbound group sessions and
+    // additionally some static number of steps:
+    //     * opening the store
+    //     * the Account
+    //     * the cross signing keys
+    //     * the tracked users
+    //     * the final save operation
```

Comment:
```suggestion
    // The total steps here include all the sessions/inbound group sessions and
    // additionally some static number of steps:
    //
    // 1. opening the store
    // 2. the Account
    // 3. the cross signing keys
    // 4. the tracked users
    // 5. the final save operation
```

Just so it's clearer this is the `5 +` in the assignment below.